pkg/server/server.go

```diff
@@ -19,57 +22,83 @@ type Service interface {
 	Run(ctx context.Context, server *Server) error
 }
 
+// Handle is a type that pairs a gRPC server to a listener
+type Handle struct {
+	Srv *grpc.Server
+	Lis net.Listener
+}
+
+// NewHandle constructs a gRPC/listener handle
+func NewHandle(srv *grpc.Server, lis net.Listener) *Handle {
+	return &Handle{Srv: srv, Lis: lis}
+}
+
+// Serve calls Serve on the gRPC server with the handle's listener
+func (h *Handle) Serve() error {
+	err := h.Srv.Serve(h.Lis)
+	if err == grpc.ErrServerStopped {
+		return nil
+	}
+	return Error.Wrap(err)
+}
+
+// Close closes the gRPC server gracefully and shuts down the listener.
+func (h *Handle) Close() error {
+	h.Srv.GracefulStop()
+	_ = h.Lis.Close()
+	return nil
+}
+
 // Server represents a bundle of services defined by a specific ID.
 // Examples of servers are the satellite, the storagenode, and the uplink.
 type Server struct {
-	lis      net.Listener
-	grpc     *grpc.Server
+	public   *Handle
+	private  *Handle
 	next     []Service
 	identity *identity.FullIdentity
 }
 
-// NewServer creates a Server out of an Identity, a net.Listener,
-// a UnaryServerInterceptor, and a set of services.
-func NewServer(opts *Options, lis net.Listener,
-	interceptor grpc.UnaryServerInterceptor, services ...Service) (
-	*Server, error) {
-	grpcOpts, err := opts.grpcOpts()
-	if err != nil {
-		return nil, err
-	}
-
-	unaryInterceptor := unaryInterceptor
-	if interceptor != nil {
-		unaryInterceptor = combineInterceptors(unaryInterceptor, interceptor)
-	}
-
+// NewServer creates a Server out of an Identity, public and private
+// gRPC handles and listeners, and a set of services. Closing a server will
+// stop the grpc servers and close the listeners.
+// A public handler is expected to be exposed to the world, whereas the private
+// handler is for inspectors and debug tools only.
+func NewServer(identity *identity.FullIdentity, public, private *Handle, services ...Service) *Server {
 	return &Server{
-		lis: lis,
-		grpc: grpc.NewServer(
-			grpc.StreamInterceptor(streamInterceptor),
-			grpc.UnaryInterceptor(unaryInterceptor),
-			grpcOpts,
-		),
+		public:   public,
+		private:  private,
 		next:     services,
-		identity: opts.Ident,
-	}, nil
+		identity: identity,
+	}
 }
 
 // Identity returns the server's identity
 func (p *Server) Identity() *identity.FullIdentity { return p.identity }
 
-// Addr returns the server's listener address
-func (p *Server) Addr() net.Addr { return p.lis.Addr() }
-
-// GRPC returns the server's gRPC handle for registration purposes
-func (p *Server) GRPC() *grpc.Server { return p.grpc }
-
 // Close shuts down the server
 func (p *Server) Close() error {
-	p.grpc.GracefulStop()
-	return nil
+	errch := make(chan error)
+	go func() {
+		errch <- Error.Wrap(p.public.Close())
+	}()
+	go func() {
+		errch <- Error.Wrap(p.private.Close())
+	}()
+	return errs.Combine(<-errch, <-errch)
```

Comment:
This should stop waiting when one of them closes with an error.

Reply:
why?

Reply:
i think it makes the most sense to wait for both here, but perhaps this is another use case for RunJointly (below)?

Reply:
Actually, maybe it should up-to a specified timeout as well. The reason to return early is so that we don't hang to unclosed connections too long. AFAIR, `GracefulStop` waits for everyone to disconnect before returning.